### ISSUE TITLE
Worklog

### ISSUE THREAD

**Issue by bmoore20** [bmoore20/habs#10]
Quickly jot down a few notes of what was accomplished after each work session. This will help keep track of the project's progress. Also include next steps.

**Comment by bmoore20** [comment]
**4.19.21**

**Accomplished:**

- Finished code for capturing seeds and updating logging, created PR, merged code (#33)
- Finished a bunch of small changes that came up during the data augmentation/transforms spike, created PR, merged code (#34)
- Researched and began process of mounting HABs GitHub repo to Google Drive

**Comment by bmoore20** [comment]
**4.21.21**

**Accomplished:**

- Researched information about how to use `device`
- Came up with clarifying questions to ask about using `device`

**Comment by bmoore20** [comment]
**4.22.21**

**Accomplished:**

- Configured all hyperparameters (optimizer, epochs, model, criterion)
  - Made helper functions in `utils`

**Next steps:**

- Abstract the `training loop` and `test loop` in _train.py_
  - Pull training and testing functionality out into their own functions (train_utils?)
  - Provide necessary hyperparameters as parameters/arguments (same ones that are passed into train.py)
- Finish device stuff
- Load mount images to Google Drive
- Install necessary requirements into Colab VM
- Run the code and fix errors/bugs that pop up

**Comment by bmoore20** [comment]
**4.2421**

**Accomplished:**

- Started to abstract the training and test loops from _train.py_

**Comment by bmoore20** [comment]
**2.25.21**

**Accomplished:**

- Finished abstraction of the training and test loops
- Submitted PR for configurations and abstractions (#39)
- Uploaded train and test images to google drive repo (_**habs_google**_)

**Next steps:**

- Finish device stuff and submit PR
- Fix up and organize **`habs_launchpad`**
  - Organize imports into one cell at the top
  - Install necessary requirements into Colab VM
  - Copy and paste training code (and other appropriate code -> initiating objects and logs) into Colab cells
- Run the code (train and evaluate model) and fix errors/bugs that pop up

**Comment by bmoore20** [comment]
**2.26.21**

**Accomplished:**

- Finished code for `device` and made PR
- Fixed up and organized **`habs_launchpad`** in Colab based on the bullet points listed above

**Next steps:**

- Need to pull down code or re-mount google drive so it contains the updated code from the new PRs (abstraction/configuration and device)

**Comment by bmoore20** [comment]
**2.27.21**

**Accomplished:**

- Read blogs and explanations on WHY validation set is needed

**Next steps:**

- Read more about validation and learn HOW it works 
- Set up **`Black`** code-formatter with HABs repo
- Fix up validation set based on Ty's suggestions in PR #39

**Comment by bmoore20** [comment]
**4.28.21**

**Accomplished:**

- Read up on how to include the code formatter `black` and the linter `flake8` with a git pre-commit hook
- Implemented the changes for adding `black` and `flake8` to the HABs repo (created PR)

**Next steps:**

- See above (minus bullet about black)

**Comment by bmoore20** [comment]
**Tomorrow:**

- Create validation dataset in train.py
- Create validation dataloader in train.py
- Create a training_lap method in training_helper.py
  - `model.train()`?
- Create a validation_lap method in training_helper.py
  - `model.eval()`?
  - No optimizer or back prop
- Write following in train.py
   ```
   for epoch in range(epoch):
       train_loss = training_lap(...)
       val_los = validation_lap(...)
       if i % 2000......
   ```
- Run in Colab (have to pull it in)

**Comment by bmoore20** [comment]
**5.31.21**

- Watched lecture on CNNs from MIT
  https://www.youtube.com/watch?v=iaSUYvmCekI

**Comment by bmoore20** [comment]
**5.31.21**

- Reviewed MIT CNN lecture and took notes
- Batch Normalization

**Comment by bmoore20** [comment]
**6.1.21**

- When I run the program in Colab with TensorBoard, I get an error `non-iterable type None` after epoch 1
- When I comment out TensorBoard stuff, whole program completes ... still gettting 0% accuracy 
- Need to get TensorBoard stuff working
  - Try moving TensorBoard calls to very bottom of notebook since nothing is writing to it right now
  - Set up a second Colab notebook (no GPU) for TensorBoard 
  - Incorporate code for log_dir -> current datetime 
- Create ResNet and Blocks

**Comment by bmoore20** [comment]
**Week of 6.22.21 - 6.29.21**

Accomplished:
- Ran the ResNet model at least once a day (all Colab could handle), experimenting with different parameters
  - Result logs can be found in https://drive.google.com/drive/u/0/folders/1aTrw2vdnLUtQl4mYsEzkWJf38e15DeeL
  - Only switched # of epochs, batch size and mag inc
  - Learn rate (1e-3), loss (cross entropy) and optimizer (adam) were all kept constant
  - Also ran Original HABs model a few times and got similar results to ResNet
  - Highest accuracy percentage was 50% ... never got higher than that 
  - I ran all of the models with a 100 mag inc (~20,000) except the last run where I switched the mag inc to 1000 -> model didn't even finish one epoch
- Looked into how to incorporate the classifying functionality from my original program into new refactored version
  - printed out results from `outputs.data` and `_, predicted = torch.max(outputs.data, 1)`
  - **All of the results were predicted to be 1 (Clear)!! -> this was a "fake" run because I used parameters that would make the run go fast to see if the printing worked ... but NEED to look into seeing if 1 is being predicted for all values in actual runs (ResNet)
- Figured out how to load in a branch from GitHub repo to a Colab notebook, and not just the Main branch
  - Allows me to experiment with different functionalities and test things out in colab without having to merge to Main 
  - Made a new Colab notebook `habs_experiment.ipynb` that runs a BRANCH (feature/experiment) from the HABs GitHub repo
  - This branch prints out tensors/arrays that have the predicted values 
  - Maybe eventually add this printing permanently in our logs ... kinda nice to see the predicted values 
- Looked to see if there were any available HAB data sets online that we could use
  - Found that there seems to already be programs similar to mine
  - Made me think about scope of project -> all of the images are not uniform
  - Have model make predictions with other objects in the images as well
  - My data is fairly uniform

Next Steps:
- Print out results for all runs and see if 1 (Clear) is being predicted for all values
- Determine why Test is only working on 8 images and NOT 10?!
- Test ResNet model not on dataset to make sure it is behaving properly 
- Think about possible next paths:
1. Need to figure out ways to improve Dataset!!!
   - How many values? Thousands? Millions?
   - Dataset from online?
   - More images from HWS professors?
   - **Transfer Learning**?!
2. Alternate to Colab?
3. What's the endgame? What functionality should I be aiming towards once we get the classification? What do users/professors want?
   - Sort images into folders? 
   - Continuous stream? 
   - Make easy to use -> Docker! 
   - Reach out to professors and see what they prefer ... now's the time

Questions:
- `outputs.data` -> is data an attribute of nn.Module?
- training loss vs. validation loss?

**Comment by bmoore20** [comment]
**7.3.21**

Accomplished:

- Started and completed PR #61 
  - experimented first with the functionality with the `feature/experiment` branch in Colab's `habs_experiment` notebook
  - removed logging completely from `training_helper.py`
  - made `evaluate` return a Tuple with 4 values -> [predictions, classifications, total, correct]
 - Merged new branch with `Main`
 - Started and completed PR #62 
   - realized that I forgot the typing for the `evaluate` method
 - Merged new branch with `Main`
 - Merged new updates to `Main` to `feature/two-classes` branch
   -  Used command: `git merge origin/main`
- Manually updated the `train` cell in `habs_launchpad`, `habs_experiment`, and `habs_two_classes` with the new log statements 
```
    predictions, classifications, total, correct = evaluate(model, test_loader)
    logger.info(f"Predicted Values: {predictions}")
    logger.info(f"Classifications: {classifications}")
```

**Comment by bmoore20** [comment]
**7.4.21**

- Update worklog for current week's work
